### PR TITLE
#1178 Fix AttributeError: 'ChatCompletionChunk' object has no attribute 'get'

### LINKED ISSUE #1178: AttributeError: 'ChatCompletionChunk' object has no attribute 'get'
**Describe the bug**
I am encountering an AttributeError while integrating LlamaIndex with Chainlit. I am trying to stream the response generated by the Chat Engine (GPT-4 and 4o) while using the LlamaIndexCallbackHandler, but when the response message is finished, I get the following error:

_AttributeError: 'ChatCompletionChunk' object has no attribute 'get'
2024-07-26 16:30:30 - 'ChatCompletionChunk' object has no attribute 'get'
[...]
File "...\llama_index\core\callbacks\base.py", line 131, in on_event_end
handler.on_event_end(event_type, payload, event_id=event_id, **kwargs)
File "...\chainlit\llama_index\callbacks.py", line 163, in on_event_end
model = raw_response.get("model", None) if raw_response else None
^^^^^^^^^^^^^^^^
[...]_

**Versions:**
LlamaIndex 0.10.58
Chainlit: 1.1.400
LLMs: GPT-4 and GPT-4o over AzureOpenAI

**Workaorund**

under chainlit/llama_index/callbacks.py/LlamaIndexCallbackHandler

I imported: 

`from openai.types.chat.chat_completion_chunk import ChatCompletionChunk`

and **added** under on_event_end:

```
if raw_response: 
    if isinstance(raw_response, ChatCompletionChunk):
        model=raw_response.model
    else:
        model = raw_response.get("model", None)
else:
    model=None
```

**instead** of this:

`model = raw_response.get("model", None) if raw_response else None`

### PR DESCRIPTION
**Pull request by dvquy13**
Fixes #1178

### CODE REVIEW

File: backend/chainlit/llama_index/callbacks.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def on_event_end(
 
             token_count = self.total_llm_token_count or None
             raw_response = response.raw if response else None
-            model = raw_response.get("model", None) if raw_response else None
+            model = raw_response.model if hasattr(raw_response, "model") else None
```
Comment (dokterbob) [inline backend/chainlit/llama_index/callbacks.py:157]: I would suggest changing this to `getattr(model, "mode", None)`.

It technically does the same thing but is more common and readable and thus maintainable.

**Review by dokterbob — CHANGES_REQUESTED** [review]
Thanks for the contrib. Please see slight suggestion here. 

Other than that, it looks good to me, from a code review standpoint.

Would be great to have feedback from more users familiar to the issue like. @logan-markewich @Tejaswgupta

File: backend/chainlit/llama_index/callbacks.py (hunk 1 above)
Comment (dvquy13) [reply backend/chainlit/llama_index/callbacks.py]: Thank you, TIL! Updated the PR

**Comment by dokterbob** [comment]
Thanks for the quick fixup. There currently is a problem with running builds, as soon as that's solved I'll re-run tests.

Mean time, could you provide code which will replicate the issue, so I can confirm it and confirm it's solution as well?

**Comment by dvquy13** [comment]
> Thanks for the quick fixup. There currently is a problem with running builds, as soon as that's solved I'll re-run tests.
> 
> Mean time, could you provide code which will replicate the issue, so I can confirm it and confirm it's solution as well?

Sure, here are the details:
<details>

<summary>Set up</summary>

Dependencies:
```
python = "3.11.9"
llama-index = "^0.10.65"
python-dotenv = "^1.0.1"
chainlit = "^1.1.402"
```

`.env` file contains: `OPENAI_API_KEY=`

</details>

<details>

<summary>Code</summary>

```
from dotenv import load_dotenv

load_dotenv()
import chainlit as cl
from chainlit import LlamaIndexCallbackHandler
from llama_index.core.agent import ReActAgent
from llama_index.core.callbacks import CallbackManager
from llama_index.core.tools import FunctionTool
from llama_index.llms.openai import OpenAI

# Ref: https://docs.llamaindex.ai/en/stable/understanding/agent/basic_agent/


def multiply(a: float, b: float) -> float:
    """Multiply two numbers and returns the product"""
    return a * b


multiply_tool = FunctionTool.from_defaults(fn=multiply)


def add(a: float, b: float) -> float:
    """Add two numbers and returns the sum"""
    return a + b


add_tool = FunctionTool.from_defaults(fn=add)


@cl.on_chat_start
async def start():
    llm = OpenAI(model="gpt-4o-mini", temperature=0)
    agent = ReActAgent.from_tools(
        [multiply_tool, add_tool],
        llm=llm,
        verbose=True,
        callback_manager=CallbackManager([LlamaIndexCallbackHandler()]),
    )
    cl.user_session.set("agent", agent)

    await cl.Message(
        author="Assistant", content="Hello! Im an AI assistant. How may I help you?"
    ).send()


@cl.on_message
async def main(message: cl.Message):
    agent = cl.user_session.get("agent")

    msg = cl.Message(content="", author="Assistant")

    res = await cl.make_async(agent.stream_chat)(message.content)

    for token in res.response_gen:
        await msg.stream_token(token)
    await msg.send()

```
</details>

<details>

  <summary>Error</summary>

```
2024-08-19 23:17:34 - Loaded .env file
2024-08-19 23:17:35 - Created default chainlit markdown file at /Users/dvq/frostmourne/tmp-chainlit-llamaindex/chainlit.md
2024-08-19 23:17:35 - Your app is available at http://localhost:8000
2024-08-19 23:18:38 - Translated markdown file for en-US not found. Defaulting to chainlit.md.
> Running step a1171073-004c-4565-b9a7-2a81eb2f135c. Step input: hi, my name is Quy
2024-08-19 23:18:47 - HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
Exception in thread Thread-1 (run):
Traceback (most recent call last):
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/chainlit/__init__.py", line 164, in with_parent_id
    await func(message)
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/test.py", line 54, in main
    for token in res.response_gen:
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/chat_engine/types.py", line 258, in response_gen
    raise self.exception
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/instrumentation/dispatcher.py", line 260, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/chat_engine/types.py", line 164, in write_response_to_history
    for chat in self.chat_stream:
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/agent/react/step.py", line 517, in gen
    yield from chat_stream
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/llms/callbacks.py", line 208, in wrapped_gen
    callback_manager.on_event_end(
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/callbacks/base.py", line 131, in on_event_end
    handler.on_event_end(event_type, payload, event_id=event_id, **kwargs)
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/chainlit/llama_index/callbacks.py", line 157, in on_event_end
    model = raw_response.get("model", None) if raw_response else None
            ^^^^^^^^^^^^^^^^
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/pydantic/main.py", line 828, in __getattr__
    raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}')
AttributeError: 'ChatCompletionChunk' object has no attribute 'get'
2024-08-19 23:18:47 - 'ChatCompletionChunk' object has no attribute 'get'
Traceback (most recent call last):
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/chainlit/utils.py", line 44, in wrapper
    return await user_function(**params_values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/chainlit/__init__.py", line 164, in with_parent_id
    await func(message)
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/test.py", line 54, in main
    for token in res.response_gen:
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/chat_engine/types.py", line 258, in response_gen
    raise self.exception
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/instrumentation/dispatcher.py", line 260, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/chat_engine/types.py", line 164, in write_response_to_history
    for chat in self.chat_stream:
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/agent/react/step.py", line 517, in gen
    yield from chat_stream
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/llms/callbacks.py", line 208, in wrapped_gen
    callback_manager.on_event_end(
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/llama_index/core/callbacks/base.py", line 131, in on_event_end
    handler.on_event_end(event_type, payload, event_id=event_id, **kwargs)
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/chainlit/llama_index/callbacks.py", line 157, in on_event_end
    model = raw_response.get("model", None) if raw_response else None
            ^^^^^^^^^^^^^^^^
  File "/Users/dvq/frostmourne/tmp-chainlit-llamaindex/.venv/lib/python3.11/site-packages/pydantic/main.py", line 828, in __getattr__
    raise AttributeError(f'{type(self).__name__!r} object has no attribute {item!r}')
AttributeError: 'ChatCompletionChunk' object has no attribute 'get'
```
</details>

**Review by dokterbob — APPROVED** [review]

**Comment by dokterbob** [comment]
Thanks @dvquy13!

**Pull request merged**